src/Compression/CompressionCodecEncrypted.h

```diff
@@ -1,104 +1,117 @@
+    * as otherwise our engines like ReplicatedMergeTree cannot
+    * deduplicate data blocks.
+    */
+class CompressionCodecEncrypted : public ICompressionCodec
+{
+public:
+    /** If a key is available, the server is supposed to
+        * invoke this static method at the startup. The codec will
+        * refuse to compress or decompress any data until that. The
+        * key can be an arbitrary octet string, but it is
+        * recommended that the key is at least 16 octets long.
+        *
+        * Note that the key is currently not guarded by a
+        * mutex. This method should be invoked no more than once.
+        */
+    CompressionCodecEncrypted();
+
+    uint8_t getMethodByte() const override;
+    void updateHash(SipHash & hash) const override;
+
+    bool isCompression() const override
     {
-    public:
-        /** If a master key is available, the server is supposed to
-          * invoke this static method at the startup. The codec will
-          * refuse to compress or decompress any data until that. The
-          * key can be an arbitrary octet string, but it is
-          * recommended that the key is at least 16 octets long.
-          *
-          * Note that the master key is currently not guarded by a
-          * mutex. This method should be invoked no more than once.
-          */
-        static void setMasterKey(const std::string_view & master_key);
+        return false;
+    }
 
-        CompressionCodecEncrypted(const std::string_view & cipher);
+    bool isGenericCompression() const override
+    {
+        return false;
+    }
 
-        uint8_t getMethodByte() const override;
-        void updateHash(SipHash & hash) const override;
+    bool isPostProcessing() const override
+    {
+        return true;
+    }
+    static void updateEncryptionKeys(const Poco::Util::AbstractConfiguration & config, const String & config_prefix);
 
-        bool isCompression() const override
-        {
-            return false;
-        }
+protected:
+    static bool tryLoadEncryptionKey(const Poco::Util::AbstractConfiguration & config, const String & config_prefix);
 
-        bool isGenericCompression() const override
-        {
-            return false;
-        }
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
 
-        bool isPostProcessing() const override
-        {
-            return true;
-        }
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override;
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override;
 
-    protected:
-        UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
+private:
+    static std::string lastErrorString();
+    static void encrypt(const std::string_view & plaintext, char * ciphertext_and_tag);
+    static void decrypt(const std::string_view & ciphertext_and_tag, char * plaintext, UInt64 key_id);
```

Comment:
Could we move those function to an anonymous namespace in `CompressionCodecEncrypted()`?